src/artifacts.js

```diff
@@ -175,8 +217,46 @@ api.declare({
   // Construct details for different storage types
   var isPublic = /^public\//.test(name);
   var details  = {};
+  let uploadId;
   switch (storageType) {
+    case 'blob':
+      // Generate the details that we'll be using.
+      details.size = input.size;
+      details.sha256 = input.sha256;
+      if (input.parts) {
+        details.parts = input.parts;
+      }
+      // TODO: These ought to be configurable
+      details.provider = 's3'; // MAKECONFIG
+      details.region = 'us-east-1'; // MAKECONFIG
+      if (input.encoding) {
+        details.encoding = input.encoding;
+      }
+
+      if (isPublic) {
+        details.bucket = 'taskcluster-blobs'; // MAKECONFIG
+      } else {
+        details.bucket = 'taskcluster-private-blobs'; // MAKECONFIG
+      }
+
+      details.key = encodeBlobKey(taskId, runId, name);
+      if (input.parts) {
+        uploadId = await this.s3Controller.initiateMultipartUpload({
```

Comment:
We are choosing to understand the S3 Rest API by doing this remote signing of requests in the first place.  I've chosen to implement this as a library without the `aws-sdk` library but rather the `aws4` signing library.  Since we need to understand how to generate the general form of an S3 rest API request, and we also need to understand how to run a request of that form, implementing the init/complete/abort calls is really trivial.  There's a lot of tests to cover these api calls.

Having this as a self-contained library also means that we can import the library already knowing that all of these methods work, and more importantly work together.  We can have testing in the queue be about testing how the queue works, and not about the low level details about how S3 uploading works.

This library also is written in a very general way which allows others in the open source community to benefit from the work we're doing here, without having to import the entire `aws-sdk` library or figure out how to use them together.

The S3 Rest API is very well documented and has a lot of information on how to use it.  It's not like we're using some sort of undocumented private API.  The `aws-sdk` library is merely a translator from Javascript to identical REST Api that I'm calling directly.  This library is not a general purpose AWS or even a general purpose S3 library.  It's a library which knows how to do uploads to S3 with a remote signing server.

If you have specific technical concerns, I am happy to address them or rethink the problem as I have done many, many times through this project.

Furthermore, this is for a new artifact type which does not have any production consumers yet.  If we land it and in testing find that this library is fundamentally unsuited to the task, I'm happy to rethink it, but this is the approach that I'm taking until there's real evidence to the contrary.

As for maintenance, I think it's highly improbable that AWS will break the S3 apis in a non-compatible way.  If they did, however, we'd be just as broken since we depend on 100% of the core functionality of the `remote-signed-s3` library working.  That doesn't change whether we call init/complete/abort through `remote-signed-s3` or `aws-sdk`.